lib/sidekiq_flow/adapters/set_storage.rb

```diff
@@ -0,0 +1,110 @@
+            connection_pool.with do |redis|
+              redis.sadd('in-progress', workflow_key) if initial
+              redis.hmset(
+                workflow_key,
+                [:klass, workflow.klass, :attrs, workflow.to_json] + workflow.tasks.map { |t| [t.klass, t.to_json] }.flatten
+              )
+            end
+          end
+      
+          def store_task(task)
+            connection_pool.with do |redis|
+              redis.hset(find_workflow_key(task.workflow_id), task.klass, task.to_json)
+            end
+            return unless workflow_succeeded?(task.workflow_id)
+            succeed_workflow(task.workflow_id)
+          end
+      
+          def find_workflow(workflow_id)
+            connection_pool.with do |redis|
+              workflow_redis_hash = redis.hgetall(find_workflow_key(workflow_id))
+              raise WorkflowNotFound if workflow_redis_hash.empty?
+              Workflow.from_redis_hash(workflow_redis_hash)
+            end
+          end
+      
+          def destroy_workflow(workflow_id)
+            connection_pool.with do |redis|
+              redis.del(find_workflow_key(workflow_id))
+            end
+          end
+      
+          def destroy_succeeded_workflows
+            workflow_keys = []
+            connection_pool.with do |redis|
+              workflow_keys = redis.smembers('finished')           
+            end
+            return if workflow_keys.empty?
+            connection_pool.with do |redis|
+              redis.del(*workflow_keys)
+              redis.srem('finished', *workflow_keys)
+            end
+          end
+      
+          """def find_workflow_keys(pattern=workflow_key_pattern)
+            connection_pool.with do |redis|
+              redis.scan_each(match: pattern, count: SCAN_COUNT).to_a
+            end
+          end"""
+
+          def find_workflow_key(workflow_id)
+            "#{configuration.namespace}.#{workflow_id}"
+          end
+      
+          def connection_pool
+            @connection_pool ||= ConnectionPool.new(size: configuration.concurrency, timeout: configuration.timeout) do
+              Redis.new(url: configuration.redis_url)
+            end
+          end
+
+          def already_started?(workflow_id)
+            workflow_key = find_workflow_key(workflow_id)
+            connection_pool.with do |redis|
+              redis.sismember('in-progress', workflow_key)
+            end
+          end
+      
+          private
+
+          def configuration
+            @configuration ||= SidekiqFlow.configuration
+          end
+
+          def generate_initial_workflow_key(workflow_id)
+            "#{configuration.namespace}.#{workflow_id}"
+          end
+
+          def workflow_succeeded?(workflow_id)
+            find_workflow(workflow_id).succeeded?
+          end
+
+          def succeed_workflow(workflow_id)
+            workflow_key = find_workflow_key(workflow_id)
+
+            return if already_succeeded?(workflow_key)
```

Comment:
we don't need this  call anymore, if the key is already in the set, redis just replies with a 0 / no-op.